### ISSUE TITLE
discussion: kUSD -> aUSD merge

### ISSUE THREAD

**Issue by NunoAlexandre** [centrifuge/centrifuge-chain#836]
## What

Our first integration with Acala involved supporting XCM transfers of `kUSD` between Altair and Karura. Now, Acala is merging `kUSD` and `aUSD` to simply `aUSD` cross network.

### Adapt XCM config
Now, we convert `kUSD` from and to `MultiLocation` like so:

***currently***
- incoming: `MultiLocation { parents: 1, X2(Parachain(2000), GeneralKey([0, 129])) }` -> `CurrencyId::KUSD`
- outgoing: `CurrencyId::KUSD` -> `MultiLocation { parents: 1, X2(Parachain(2000), GeneralKey([0, 129])) }`

`aUSD` in `Kusama` will use the same `MultiLocation` that was used for `kUSD`, stated above. So, in theory, we just have to change the `CurrencyId` variant we convert said `MultiLocation` from and to:

***becomes***
- incoming: `MultiLocation { parents: 1, X2(Parachain(2000), GeneralKey([0, 129])) }` -> `CurrencyId::AUSD`
- outgoing: `CurrencyId::AUSD` -> `MultiLocation { parents: 1, X2(Parachain(2000), GeneralKey([0, 129])) }`

### Migrate balances

As the XCM config will move to use `AUSD` and not `KUSD`, users that hold balance for `CurrencyId::KUSD` will need to be able to see those funds as AUSD and be able to transfer them through XCM as such.

## Proposal

My idea initially when dropping [`CurrencyId::USD` in favour of `CurrencyId::AUSD`](https://github.com/centrifuge/centrifuge-chain/pull/811/files) as so that we would already have the `CurrencyId` variant we would need to transfer funds to and to migrate to at the XCM config level.

However, now with the introduction of `asset-registry`, we have a new option. So we can choose from:

### Option 1 (using `asset-registry`)

1. Drop `CurrencyId::AUSD`
2. Register `AUSD` as a `CurrencyId::ForeignAsset` in the asset-registry
3. Drop all `KUSD`-related XCM logic
    automatically, `AUSD` XCM transfers will start being handled by the new XCM logic that looks at the `asset-registry`
4. Migrate all `KUSD` balances 
   4.a from `CurrencyId::KUSD` to `CurrencyId::ForeignAsset(<ausd_foreign_asset_id>)`; OR
   4.b we announce the change to the community and let the few KUSD holders run that transfer themselves
5. Drop `CurrencyId::KUSD`
6. Adapt all tests and benchmarks from using `CurrencyId::AUSD` to registering AUSD 

_Advantages_
- The more we move to the asset-registry, less specific logic we have lingering around for specific tokens. This isn't a substantial burden tho.

_Disadvantages_
- It makes testing and mocking harder since we wouldn't have a stable asset available at compile time and would need to register such asset into the asset-registry at runtime.
 

### Option 2 (using `CurrencyId::AUSD`)

1. Adapt all `KUSD`-related XCM logic to use `CurrencyId::AUSD`
    See `Changes` > `becomes` above
2. Migrate all `KUSD` balances 
   4.a from `CurrencyId::KUSD` to `CurrencyId::AUSD`; OR
   4.b we announce the change to the community and let the few KUSD holders run that transfer themselves
3. Drop `CurrencyId::KUSD` variant
4. We would still register `CurrencyId::AUSD` in the asset-registry for the sake of providing an API to query it's metadata

_Advantages_
- Simpler, less impactful change
- Potentially handy to have a stable asset available at compile time

_Disadvantages_
- We keep specific XCM logic for AUSD; which, again, is really not much of a burden anyway

**Comment by NunoAlexandre** [comment]
@offerijns @branan @mikiquantum I'd love to get your thoughts on this.

I have a slight preference for Option 2 because of the reduced impact on the repo.

🤯  **PLUS**, I jus realised this, if we choose this option I am pretty certain that we won't need to concern ourselves with the balances migration step because by dropping `CurrencyId::KUSD`, `CurrencyId::AUSD` will automatically be encoded to the value that `CurrencyId::KUSD` was, so the migration is essentially automatically done for free. I need to verify this tho!

In any case, we need to find a way of not breaking the balances when messing around with `CurrencyId` variants. Pehraps we need to implement a custom `Encode`/`Decode` impl for `CurrencyId`.

**Comment by mikiquantum** [comment]
@NunoAlexandre thanks for the write up. If internally the balance is represented by [0,129] array then we prob do not need a migration. 
In any case, I would focus on making sure we do not maintain both ways, even for a short period of time, since we, most likely, don't have much balance in either one, I would run an internal migration within the designed runtime upgrade to move these balances (if needed), and not let the users do that, since that will take more time. 
I think for now option 2 makes sense as well, later on we can think about consolidating all foreign assets in the registry. 

Can you explain what do you mean about the `encode/decode` if the underlaying currencyId is not changed?

**Comment by NunoAlexandre** [comment]
> @NunoAlexandre thanks for the write up. 

@mikiquantum Thanks for taking the time to give me feedback 👌 

---

> If internally the balance is represented by [0,129] array 

That's not the case. `[0, 129]` is aUSD's XCM MultiLocation GeneralKey value, it's not the value used in Altair for indexing kusd/ausd balances.

`OrmlTokens` uses the encoded `CurrencyId` as the key for indexing balances, and that value is essentially derived by the index of the specific `CurrencyId` variant in the enum. So my idea about being able to drop the `KUSD` variant and automatically having this migration done for free is that since the `AUSD` variant comes right after the `KUSD`, by dropping the former, the latter will have the same encoding value and therefore no migration is needed.

See https://github.com/centrifuge/centrifuge-chain/pull/837/files where I test both the value to which each variant encodes to at the moment, and how each variant of a copy of `CurrencyId` without `KUSD` would differ.

---
> Can you explain what do you mean about the `encode/decode` if the underlaying currencyId is not changed?

Sure! So, as we see in this very case, if we drop or move a variant of `CurrencyId` around, that will change what each variant will `encode` to. This would happen silently, that's why I started the PR just liked above to have the sanity checks. My suggestion on that comment you quoted as to say that we could impl `Encode` and `Decode` for `CurrencyId` ourselves and thus be explicit about how each variant is encoded/decoded.

To avoid that from happening, Acala for example marks their `enum TokenSymbol` with `#[repr(u8)]` and attributes a static value to each variant: https://github.com/AcalaNetwork/Acala/blob/67c72c9305833c61c22987495ac3c71170457df5/primitives/src/currency.rs#L202-L230 (That specific approach doesn't work for us because our tranche and foreign assets need a more sophisticated encoding logic that takes their inner values into account)

**Comment by NunoAlexandre** [comment]
@offerijns @mikiquantum @branan

tl;dr - I am dropping `CurrencyId::KUSD` and leave `CurrencyId::AUSD`. No migration is needed.

I have tested locally that dropping the `CurrencyId::KUSD` has `CurrencyId::AUSD` taking over `KUSD`'s balances due to the encoding of the `CurrencyId` variants.

By dropping `Currency::KUSD`, we are left with:
- just `CurrencyId::AUSD` with a zero-cost balance migration, 
- a variant that works both for Altair on Kusama and Centrifuge on Polkadot
- an up-to-date named currency variant
- a stable coin variant available at compilation time

I personally don't see the benefit of dropping both variants and make them `foreignAsset`s and operate a balance migration, but shall any of you feel different about this please let me know.

### Screenshots of the process of verifying that `AUSD` took over a `KUSD` balance:

##### 1.a Check `KUSD` balance `0`
![1 a kUSD](https://user-images.githubusercontent.com/7075260/174771864-7e9465f0-ea23-4e69-b8a8-47c9fb659ef1.png)

##### 1.b Check `AUSD` balance `0`
![1 b aUSD](https://user-images.githubusercontent.com/7075260/174771878-3709a3d2-125a-4d28-84ed-7e6174d5b3af.png)

##### 2. Set. `KUSD` balance to 123
![2  set kUSD balance](https://user-images.githubusercontent.com/7075260/174771882-9c0e1dfb-9bb0-40fe-9b8e-f8672c3c21bd.png)

##### 3.a Verify `KUSD` balance is 123
![3  kUSD 123](https://user-images.githubusercontent.com/7075260/174771884-4135bfad-b480-4f3b-b7d5-709dc42c9036.png)

##### 3.b Verify `AUSD` balance is still 0
![3 b aUSD 0](https://user-images.githubusercontent.com/7075260/174771888-0abbfbab-477c-4f04-8a03-8eae7c20d75b.png)

#### --- Ran the Runtime Upgrade that dropped `KUSD` ----

##### 3. Verify `AUSD` balance is 123
![4  auSD 123](https://user-images.githubusercontent.com/7075260/174771890-e2565eb9-89d6-49b2-bee2-462ee0adb3d8.png)
